sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -4229,6 +4228,22 @@ static pi_result cleanupAfterEvent(pi_event Event) {
         if (it == Queue->ZeCommandListFenceMap.end()) {
           die("Missing command-list completition fence");
         }
+
+        // Workaround for VM_BIND mode.
+        // Make sure that the command-list doing memcpy is reset before non-USM
+        // host memory potentially involved in the memcpy is freed.
+        //
+        // NOTE: it is valid to wait for the fence here as long as we aren't
+        // doing batching on the involved command-list. Today memcpy goes by
+        // itself in a command list.
+        //
+        // TODO: this will unnecessarily(?) wait for non-USM memory buffers too,
+        // so we might need to add a new command type to differentiate.
+        //
+        if (Event->CommandType == PI_COMMAND_TYPE_MEM_BUFFER_COPY) {
```

Comment:
This does not address the scenario of a chain of command lists, if the copy list is in the middle of such chain, right?

Reply:
> This does not address the scenario of a chain of command lists, if the copy list is in the middle of such chain, right?

I think it does.  Because when we call cleanupAfterEvent, this recursively calls cleanupAfterEvent on any events that the current event had in its event wait list.

Reply:
Actually, not truly recursively, but in a loop that you can think of as implementing that recursion.

Reply:
@kbsmith-intel  where is the event wait list created? I mean, does that wait list include events passed as waitevents on the same command, something like:
```cpp
zeAppendLaunchKernel(signalEvent, countOfWaitEvents, waitEventsList);
```
or every single event in the chain of lists/commands?

Reply:
Each event has a command that is associate with it.  Each event also has a list of events that the command associated with it had in its waitEventsList.  That list is created by createAndRetainPiZeEventList routine.  That routine returns the list, which gets associated with the event once the event gets created.  By walking this list of events recursively, you can construct a full list of every single event that the event/command are dependent on.

Reply:
great, cool.